Added recipe for htop

keuv-grvl opened the pull request.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/htop```) and found some lint.

Here's what I've got...


For **recipes/htop**:

 * The recipe could do with some maintainers listed in the `extra/recipe-maintainers` section.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/htop```) and found it was in an excellent condition.

jakirkham commented:
Seems like a nice recipe to have. What is the reason for closing? Please let us know if there is anything we can do.

Normally I use [`glances`]( http://nicolargo.github.io/glances/ ) for this as it works everywhere including Windows. Though `htop` is still definitely a nice utility to have.

keuv-grvl commented:
Ow! Closing the PR was non intended.

Nonetheless, my local test (a simple `conda build recipes/htop && conda install htop --use-local` in an empty conda-env) do not work if conda get `ncurses` from conda-forge channel (5.9-9). But it works with the default channel ([ncurses 5.9-8](https://github.com/conda-forge/ncurses-feedstock/commit/a03ea33d55e4939c9df4b007c507fbc046127cf5)).
Any idea why?

jakirkham commented:
Ok, cool. Just wanted to make sure you hadn't lost hope. 😄

Hmm...so `ncurses` build number `9` of `5.9` dropped the argument `--with-terminfo-dirs`. Please see PR ( https://github.com/conda-forge/ncurses-feedstock/pull/19 ) for details. That had to be dropped as it caused issues ( https://github.com/conda-forge/ncurses-feedstock/issues/17 ) for other users. We could try to bring it back with a list of directories to search.

Could you please provide info about your system? Guessing Linux of some kind?

keuv-grvl commented:
I use Ubuntu 16.04.1 x86_64 with conda 4.2.6.

I'm not familiar with `ncurses`, but maybe [this](https://lists.gnu.org/archive/html/bug-ncurses/2011-06/msg00020.html) and [this](https://github.com/conda/conda-recipes/pull/223/files) could help.

jakirkham commented:
Yep, that seems right. Could you please submit a PR?

jakirkham commented:
Yep, that seems right. Could you please submit a PR to the [feedstock]( https://github.com/conda-forge/ncurses-feedstock )?

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+
+source:
+  git_url: https://github.com/hishamhm/htop.git
+  git_rev: {{ version }}
```
Could we please switch this to a tarball instead?

```yaml
  fn: {{ name }}-{{ version }}.tar.gz
  url: https://github.com/hishamhm/htop/archive/{{ version }}.tar.gz
  sha256: <some checksum>
```

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+requirements:
+  build:
+    - automake
+    - gcc
```
Please drop `gcc` as we already include the compilers on the VMs.

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  skip: true  # [win]
+
+requirements:
+  build:
```
Please add `toolchain` to properly configure the VM compilers.

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+    - automake
+    - gcc
+    - libtool
+    - ncurses
```
Please pin to `5.9*` to match our [global pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/70036d169783f7d6844bfc1ff1b360dba5adfa9f/scripts/pin_the_slow_way.py#L45 ).

jakirkham replied:
Also shouldn't this be a run time dependency too?

keuv-grvl replied:
- I'm not sure if I pinned it right...
- Now added to run time dependency

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+
+about:
+  home: https://github.com/hishamhm/htop/
+  license: GNU General Public License, version 2 (GPL-2.0)
```
Please shorten to `GPL-2` or similar.

jakirkham replied:
Also please include the license file by adding `license_file: COPYING`.

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+
+extra:
+  recipe-maintainers:
+    - keuv-grvl
```
nit: Please add terminal newline.

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  build:
+    - automake
+    - gcc
+    - libtool
```
Should we add `pkg-config`? Seems like we can't find the headers for `ncurses`. So this may help.

keuv-grvl replied:
It did not change anything on my system, so I did not add it. But let met know if I have to.

jakirkham commented on recipes/htop/build.sh:
```diff
@@ -0,0 +1,6 @@
+set -euo pipefail
+./autogen.sh
+./configure --prefix=$PREFIX
+make
```
Is there a `make check` that we can run?

keuv-grvl replied:
There is one, but it does not seem to check something. I added it anyway.

jakirkham commented:
Given the changes mentioned above and your PR ( https://github.com/conda-forge/ncurses-feedstock/pull/23 ), I was able to build `htop` an run it locally. So I think we are on the right path. 😄

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.

keuv-grvl commented:
Thank your @jakirkham for your review!

I added your suggestions to the current PR, and it works on my machine using:
``` bash
conda build recipes/htop
conda search htop --use-local
conda install htop --use-local
```
Let me know if you can reproduce it.
Also, I did not use [conda-forge/ncurses-feedstock#23](https://github.com/conda-forge/ncurses-feedstock/pull/23) but the current `ncurses` from conda-forge (`5.9 - 9`).

The odd thing is: I dropped `conda-forge` from my channels (using `conda config --remove-key channels`) to use only channel `defaults`, then tried to build the package, and it worked (I had to remove `CFLAGS` and `toolchain`). These packages were used as dependencies:
- `libgcc-5.2.0 - 0`
- `libtool-2.4.2 - 0`
- `m4-1.4.17 - 0`
- `autoconf-2.69 - 0`
- `ncurses-5.9 - 8`

So I guess `ncurses-5.9 - 9` from the conda-forge is "breaking" something, but I can't figure out how precisely...
Anyway, we are making some progress :smiley:

keuv-grvl commented on recipes/htop/build.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+set -euo pipefail
```
Done

jakirkham commented on recipes/htop/COPYING:
```diff
@@ -0,0 +1,355 @@
+        GNU GENERAL PUBLIC LICENSE
```
We shouldn't need this right? We are just grabbing it from the source code.

keuv-grvl replied:
So which license file should I add?

jakirkham replied:
We shouldn't need to add one.

The feedstock will get a BSD 3-Clause license as it is the standard.

Adding the `license_file` portion in the recipe was the only thing we needed to do. However, it pulls the license file from the `htop`'s source. So it isn't using this file either.

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+    - toolchain
+    - automake
+    - libtool
+    - ncurses >=5.9
```
This should be `ncurses 5.9*`. Otherwise this will break when `ncurses 6.0` is released here.

keuv-grvl replied:
Done

jakirkham commented on recipes/htop/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+    - ncurses >=5.9
+
+  run:
+    - ncurses >=5.9
```
Same as above.

keuv-grvl replied:
Done

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/htop```) and found it was in an excellent condition.

jakirkham commented:
To build, we normally use a Docker container. This should be easy to do locally as long as you have [Docker installed]( https://docs.docker.com/engine/installation/ ). To get the build artifacts out, you may need to use PR ( https://github.com/conda-forge/staged-recipes/pull/1508 ). This will drop the package in `build_artefacts/linux-64`. Please try installing the package on your Linux machine with `conda-forge` enabled and let me know how that goes.

jakirkham commented:
So I built this fine in the Docker container and was able to use it successfully in our container and Continuum's. Not sure what is causing you build issues locally, but would recommend using our Docker container and seeing if the resulting package will work on your OS.

keuv-grvl commented:
I successfully built `htop` using the Docker container. I also could install and run `htop` within the Docker container.
Then [I added](https://github.com/conda/conda/issues/424) the `build_artefacts/` folder to my channels in the `~/.condarc` file and I could install `htop` from this local channel.

Everything seems to work fine :)

So is [conda-forge/ncurses-feedstock#23](https://github.com/conda-forge/ncurses-feedstock/pull/23) still relevant?

jakirkham commented:
Great! Glad this works for you. 😄

We can take a closer look at the build issues you encountered outside the Docker image in an issue at the feedstock. Just make sure to ping me as I'm often unsubscribed from feedstocks to keep notifications manageable.

> So is [conda-forge/ncurses-feedstock#23](https://github.com/conda-forge/ncurses-feedstock/pull/23) still relevant?

No, it sounds like it isn't. Would you mind closing it and making a note about this?

jakirkham commented:
Could you please revert/remove commit ( https://github.com/conda-forge/staged-recipes/pull/1598/commits/c3ef0b816ef9f4c6dd3d1497061e19076988bfbc ) from the PR?

jakirkham commented:
Thanks @keuv-grvl.

jakirkham commented:
Soon you should get an email that invite you to join conda-forge, @keuv-grvl. Once accepted you will be added to a team with the same name as this recipe. Those will give you permissions on the feedstock (repo) for this recipe.

Make sure when proposing any change that you go through the typical GitHub workflow of forking the feedstock and making changes in your fork that you PR back. Once merged CIs will build and deploy any changes you make. Please let us know if you have any questions and welcome to conda-forge.

keuv-grvl commented:
`htop` was my first recipe.  So thank you @jakirkham for your precious help.
